File: solutions/erlang/diamond/1/src/diamond.erl
-module(diamond).

-export([rows/1]).

rows([$A]) -> ["A"];
rows([Letter]) when Letter > $A, Letter =< $Z -> 
    Letters = lists:reverse(lists:seq($A, Letter)),
    LettersWithPaddings = lists:reverse(with_index(Letters)),
    DiamondStructure = mirror(LettersWithPaddings),
    lists:map(fun row/1, DiamondStructure).

row({Letter, Padding}) ->
    Pad = string:copies(" ", Padding),
    Pad ++ diamond(Letter) ++ Pad.

diamond($A) -> "A";
diamond(Letter) -> 
    MiddleSpaces = (Letter - $B) * 2 + 1,
    [Letter] ++ string:copies(" ", MiddleSpaces) ++ [Letter].

with_index(List) -> with_index(List, 0, []).

with_index([Item | List], Index, Result) -> 
    with_index(List, Index + 1, [{Item, Index} | Result]);
with_index([], _, Result) -> lists:reverse(Result).

mirror([]) -> [];
mirror(List) -> 
    [_ | Reversed] = lists:reverse(List),
    List ++ Reversed.

File: solutions/elixir/diamond/1/lib/diamond.ex
defmodule Diamond do
  @doc """
   This generates a sequence that is defined by number of spaces inside of diamond
   for given letter (zero-indexed, but we omit "A", because it produces a spec-case, so this starts from letter "B")
  """
  @letter_sizes Enum.to_list(1..49//2)

  @doc """
  Given a letter, it prints a diamond starting with 'A',
  with the supplied letter at the widest point.
  """
  @spec build_shape(char) :: String.t()
  def build_shape(letter) when letter in ?A..?Z, 
    do: letter..?A
      |> Enum.with_index()
      |> Enum.reverse()
      |> mirror()
      |> Enum.map(fn {letter, padding} -> 
        pad = String.duplicate(" ", padding)
        pad <> diamond(letter) <> pad <> "\n"
      end)
      |> Enum.join()

  def diamond(?A), do: "A"
  Enum.each(Enum.with_index(@letter_sizes, 1), fn {size, letter_index} ->
    letter = ?A + letter_index
    def diamond(unquote(letter)), 
      do: <<unquote(letter), String.duplicate(" ", unquote(size))::binary, unquote(letter)>>
  end)

  defp mirror(list) do
    [_ | reversed] = Enum.reverse(list)
    list ++ reversed
  end
end
